python/iceberg/core/filesystem/s3_filesystem_wrapper.py

```diff
@@ -0,0 +1,283 @@
+        is_dir = False
+        st = dict()
+        try:
+            st = S3FileSystemWrapper.info(S3FileSystemWrapper.fix_s3_path(path), refresh=True)
+        except RuntimeError:
+            # must be a directory or subsequent du will fail
+            du = S3FileSystemWrapper.du(S3FileSystemWrapper.fix_s3_path(path), refresh=True)
+            if len(du) > 0:
+                is_dir = True
+            length = 0
+            for key, val in du.items():
+                length += val
+
+        return FileStatus(path=path, length=st.get("Size", length), is_dir=is_dir,
+                          blocksize=None, modification_time=st.get("LastModified"), access_time=None,
+                          permission=None, owner=None, group=None)
+
+    @staticmethod
+    def full_file(url, role_arn="default", use_requests=True):
+        if use_requests:
+            url = S3FileSystemWrapper.get_presigned_urls([url])[0]
+            stream = requests.get(url, stream=True).content
+        else:
+            stream = S3File(url, role_arn).read()
+
+        return io.BytesIO(stream)
+
+    @staticmethod
+    def get_presigned_urls(urls, role_arn="default"):
+        signed_urls = []
+        for url in urls:
+            parsed_url = urlparse(S3FileSystemWrapper.fix_s3_path(url))
+            signed_urls.append(get_s3(role_arn, "client").generate_presigned_url(ClientMethod='get_object',
+                                                                                 Params={'Bucket': parsed_url.netloc,
+                                                                                         'Key': parsed_url.path[1:]}))
+        return signed_urls
+
+    @staticmethod
+    def du(url):
+        raise NotImplementedError()
+
+    @staticmethod
+    def info(url):
+        raise NotImplementedError()
+
+    @staticmethod
+    def fix_s3_path(path):
+        if path.startswith("s3n"):
+            path = path.replace("s3n://", "s3://")
+        return path
+
+
+class S3File(object):
+    MAX_CHUNK_SIZE = 4 * 1048576
+    MIN_CHUNK_SIZE = 2 * 65536
+
+    def __init__(self, path, role_arn="default", localroot="."):
+        self.path = path
+        self.parsed_url = urlparse(path)
+        self.curr_pos = 0
+        self.obj = (get_s3(role_arn=role_arn)
+                    .Object(bucket_name=self.parsed_url.netloc,
+                            key=self.parsed_url.path[1:]))
+        self.size = self.obj.content_length
+        self.isatty = False
+        self.closed = False
+
+        self.buffer_remote_reads = True
+
+        self.curr_buffer = None
+        self.curr_buffer_start = -1
+        self.curr_buffer_end = -1
+        self.buffer_reads = 0
+        self.buffer_hits = 0
+
+        # materialized optimization
+        self.materialized = False
+        self.local_path = None
+        self.local_fo = None
+        self.localroot = localroot
+
+        self.chunk_size = self.MAX_CHUNK_SIZE
+
+    def materialize(self):
```

Comment:
It would help to document what an operation like this means as it's not really clear what the intent is.

Reply:
this is an optimization that pulls the file locally and then operates on it as a local file handle.